BUG: np.gcd returns negative result for INT_MIN input on s390x

### Describe the issue:

np.gcd returns a negative value when one of the inputs is the minimum value of a signed integer type on s390x. The signed-to-unsigned absolute value conversion in npy_gcd (npy_math_internal.h.src) uses `a < 0 ? -a : a`, which is undefined behavior when `a = INT_MIN`.

Found it when running tests on a RHEL9/CentOS Stream 9 machine with Python3.14 and GCC 11.5. I wasn't able to reproduce it on other archs or even later GCC versions.

```

___________________ TestRationalFunctions.test_gcd_overflow ____________________

self = <test_umath.TestRationalFunctions object at 0x3ff66a32690>

    def test_gcd_overflow(self):
        for dtype in (np.int32, np.int64):
            # verify that we don't overflow when taking abs(x)
            # not relevant for lcm, where the result is unrepresentable anyway
            a = dtype(np.iinfo(dtype).min)  # negative power of two
            q = -(a // 4)
>           assert_equal(np.gcd(a,  q * 3), q)
E           AssertionError: 
E           Items are not equal:
E            ACTUAL: np.int32(-536870912)
E            DESIRED: np.int32(536870912)

a          = np.int32(-2147483648)
dtype      = <class 'numpy.int32'>
q          = np.int32(536870912)
self       = <test_umath.TestRationalFunctions object at 0x3ff66a32690>
```
This seems to be fixed when using -fwrapv


### Reproduce the code example:

```python
import numpy as np
a = np.int32(np.iinfo(np.int32).min)
q = -(a // 4)
print(np.gcd(a, q * 3))

That should produce 536870912, instead it produces -536870912
```

### Error message:

```shell

```

### Python and NumPy Versions:

Numpy 2.3.4 (but also reproduced on latest versions
Python 3.14
GCC 11.5.0
CFLAGS: -O2 -march=z14

### Runtime Environment:

[{'numpy_version': '2.3.4',
  'python': '3.14.3 (main, Feb  4 2026, 00:00:00) [GCC 11.5.0 20240719 (Red '
            'Hat 11.5.0-14)]',
  'uname': uname_result(system='Linux', node='removed', release='5.14.0-697.el9.s390x', version='#1 SMP Wed Apr 22 08:43:20 UTC 2026', machine='s390x')},
 {'simd_extensions': {'baseline': [],
                      'found': ['VX', 'VXE', 'VXE2'],
                      'not_found': []}},
 {'available_backends': ['NETLIB', 'OPENBLAS-OPENMP'],
  'current_backend': 'OPENBLAS-OPENMP',
  'filepath': '/usr/lib64/libflexiblas.so.3.0',
  'internal_api': 'flexiblas',
  'loaded_backends': ['OPENBLAS-OPENMP'],
  'num_threads': 4,
  'prefix': 'libflexiblas',
  'user_api': 'blas',
  'version': '3.0.4'},
 {'architecture': 'Z14',
  'filepath': '/usr/lib64/libopenblaso-r0.3.29.so',
  'internal_api': 'openblas',
  'num_threads': 4,
  'prefix': 'libopenblas',
  'threading_layer': 'openmp',
  'user_api': 'blas',
  'version': '0.3.29'},
 {'filepath': '/usr/lib64/libgomp.so.1.0.0',
  'internal_api': 'openmp',
  'num_threads': 4,
  'prefix': 'libgomp',
  'user_api': 'openmp',
  'version': None}]


### How does this issue affect you or how did you find it:

_No response_

--- a/numpy/_core/src/npymath/npy_math_internal.h.src
+++ b/numpy/_core/src/npymath/npy_math_internal.h.src
@@ -650,12 +650,24 @@ npy_lcm@c@(@type@ a, @type@ b)
  *
  * #type = (npy_int, npy_long, npy_longlong)*2#
  * #c = (,l,ll)*2#
+ * #utype = (npy_uint, npy_ulong, npy_ulonglong)*2#
  * #func=gcd*3,lcm*3#
  */
 NPY_INPLACE @type@
 npy_@func@@c@(@type@ a, @type@ b)
 {
+    #if defined(__s390__) || defined(__s390x__)
+    /*
+     * On s390x, negating INT_MIN results in a negative value due to how
+     * the architecture handles signed overflow. Cast to unsigned first
+     * to avoid this issue.git issue #31359
+     */
+    @utype@ a_abs = a < 0 ? (@utype@)(0) - (@utype@)a : (@utype@)a;
+    @utype@ b_abs = b < 0 ? (@utype@)(0) - (@utype@)b : (@utype@)b;
+    return (@type@)npy_@func@u@c@(a_abs, b_abs);
+#else
     return npy_@func@u@c@(a < 0 ? -a : a, b < 0 ? -b : b);
+#endif
 }
 /**end repeat**/
 

--- a/numpy/_core/src/npymath/npy_math_internal.h.src
+++ b/numpy/_core/src/npymath/npy_math_internal.h.src
@@ -656,18 +656,15 @@ npy_lcm@c@(@type@ a, @type@ b)
 NPY_INPLACE @type@
 npy_@func@@c@(@type@ a, @type@ b)
 {
-    #if defined(__s390__) || defined(__s390x__)
     /*
-     * On s390x, negating INT_MIN results in a negative value due to how
-     * the architecture handles signed overflow. Cast to unsigned first
-     * to avoid this issue.git issue #31359
+     * Cast to unsigned to avoid overflow when negating minimum signed value.
+     * For negative values, cast to unsigned first, then apply negation in
+     * unsigned arithmetic to avoid undefined behavior.
+     * This ensures correct behavior across all architectures. #31359
      */
     @utype@ a_abs = a < 0 ? (@utype@)(0) - (@utype@)a : (@utype@)a;
     @utype@ b_abs = b < 0 ? (@utype@)(0) - (@utype@)b : (@utype@)b;
     return (@type@)npy_@func@u@c@(a_abs, b_abs);
-#else
-    return npy_@func@u@c@(a < 0 ? -a : a, b < 0 ? -b : b);
-#endif
 }
 /**end repeat**/
 

--- a/numpy/_core/tests/test_umath.py
+++ b/numpy/_core/tests/test_umath.py
@@ -4300,10 +4300,14 @@ class TestRationalFunctions:
         # verify that we don't overflow when taking abs(x) for INT_MIN
         # this was undefined behavior that manifested on s390x with GCC 11.5
         for dtype in (np.int8, np.int16, np.int32, np.int64):
-            a = dtype(np.iinfo(dtype).min)  # negative power of two
+            a = dtype(np.iinfo(dtype).min)  # INT_MIN
             q = -(a // 4)
+            # Test with INT_MIN as first argument
             assert_equal(np.gcd(a,  q * 3), q)
             assert_equal(np.gcd(a, -q * 3), q)
+            # Test with INT_MIN as second argument
+            assert_equal(np.gcd(q * 3,  a), q)
+            assert_equal(np.gcd(-q * 3, a), q)
 
     def test_decimal(self):
         from decimal import Decimal

--- a/numpy/_core/tests/test_umath.py
+++ b/numpy/_core/tests/test_umath.py
@@ -4297,9 +4297,9 @@ class TestRationalFunctions:
         assert_equal(np.lcm(a, b), 10 * big)
 
     def test_gcd_overflow(self):
-        for dtype in (np.int32, np.int64):
-            # verify that we don't overflow when taking abs(x)
-            # not relevant for lcm, where the result is unrepresentable anyway
+        # verify that we don't overflow when taking abs(x) for INT_MIN
+        # this was undefined behavior that manifested on s390x with GCC 11.5
+        for dtype in (np.int8, np.int16, np.int32, np.int64):
             a = dtype(np.iinfo(dtype).min)  # negative power of two
             q = -(a // 4)
             assert_equal(np.gcd(a,  q * 3), q)

